### PR TITLE
[Issue 381 & 379] code of conduct checkbox, validation, profanity checker, and successful submission

### PR DESCRIPTION
**Pull request by daaimah123**
### 📝 Description
We don't want the user to be able to submit the form and their payment info if profanity is detected, we also want submission to acknowledge commitment to the code of conduct, handle form validation with clear indication of what is happening, and a successful submission view to remove the form. 

### 🔂 Changes Made
- Function to check text for profanity using PurgoMalum API, with focus on the first invalid field.
- Create a mapping of field IDs to readable labels.
- Checking for profanity error indication

### ⚙️ Related Issue
- Issue Number: #379 & #381

### 🍏 Type of Change
- New feature(s)

### 🎁 Acceptance Criteria
- Checks all user text input values for profanity in strings
- When profanity, notify user and prevent submission from going through and notify the user.
- Check for credit card information is entered, notifies user if missing
- Disabled submission button if code of conduct or credit card info not entered/selected
- Validation statements for missing required field inputs

### 🧪 How to test
- Ensure you have .env/config.ini variables. 
- Run instructions under [Square Testing in the README](https://github.com/Techtonica/techtonica.org?tab=readme-ov-file#square-testing)
  ```
  pip install pyopenssl
  FLASK_DEBUG=1 FLASK_APP=main_site.py FLASK_RUN_CERT=adhoc flask run
  ```
- Try to submit the form in the following ways
  - Input strings with combination of profanity and regular words to see if the form allows for submission (you should not be unable to submit, you should see either a red indication telling you to enter credit card information or identifying profane language)
  - Enter the [test card details](https://developer.squareup.com/docs/devtools/sandbox/payments) with all required fields entered, no profanity, with the checkbox selected (you should see a processing payment validation statement and get a successful screen)
  - Attempt to submit the form without the checkbox selected (you should be unable to)
  - Attempt to submit the form with the checkbox selected and no fields completed or selected (you should be unable to)
  - Attempt to submit the form with the checkbox selected, [test card details](https://developer.squareup.com/docs/devtools/sandbox/payments) entered, and no other required fields selected or input (you should be unable to, you should be prompted to enter the required field input)

### CODE REVIEW

**Comment by MichelleGlauser** [comment]
Let's change that text to "You agree that what you are submitting in this form agrees to Techtonica's Code of Conduct."

**Comment by kaylahrose** [comment]
If the action we want the user to take to provide consent is clicking the submit form button, we should remove the checkbox and keep the text. Consenting via form submission is implicit consent. However, if the action we would like the user to perform is checking the box, we should change the language to "By checking this box, you are agreeing to our Code of Conduct" or something like that. That is explicit consent, which is preferred.

File: templates/job-form.html [hunk 1]
```diff
@@ -93,6 +93,20 @@ <h3>Job Listing Form</h3>
...
+          <label for="code-of-conduct-checkbox">
+            You agree to 
+            <a id="code-of-conduct-checkbox-link" href="{{ url_for('render_conduct_page') }}"
+              title="Code of Conduct"
+              target="_blank"
+              rel="noopener noreferrer"
+              >
+              Techtonica's Code of Conduct
+            </a>
+            by submitting this form.
```
Comment (kaylahrose) [inline templates/job-form.html:107]: If the action we want the user to take to provide consent is clicking the submit form button, we should remove the checkbox and keep the text. Consenting via form submission is implicit consent. However, if the action we would like the user to perform is checking the box, we should change the language to "By checking this box, you are agreeing to our Code of Conduct" or something like that. That is explicit consent, which is preferred.

File: templates/job-form.html [hunk 2]
```diff
@@ -93,6 +93,20 @@ <h3>Job Listing Form</h3>
...
         <div id="card-container"></div>
+        <div class="checkbox-container">
+          <input type="checkbox" id="code-of-conduct-checkbox">
+          <label for="code-of-conduct-checkbox">
+            You agree to 
+            <a id="code-of-conduct-checkbox-link" href="{{ url_for('render_conduct_page') }}"
+              title="Code of Conduct"
+              target="_blank"
+              rel="noopener noreferrer"
+              >
```
Comment (kaylahrose) [inline templates/job-form.html:104]: Can you please update the link style match the [style guide](https://www.dropbox.com/scl/fo/sb8ppa0qxuuott515euei/AEG3sWxuiJ8JfU36n35H7E0?e=2&preview=Techtonica_UI_design_kit_rev1.pdf&rlkey=kasmils5rmaqv199ggoh2us5p&dl=0)?

File: templates/job-form.html [hunk 3]
```diff
@@ -93,6 +93,20 @@ <h3>Job Listing Form</h3>
         </div> -->
         <h3>Payment Info</h3>
         <div id="card-container"></div>
+        <div class="checkbox-container">
+          <input type="checkbox" id="code-of-conduct-checkbox">
```
Comment (kaylahrose) [inline templates/job-form.html:97]: Can you please separate the checkbox from the text like the photo attached
Screenshot 2025-02-28 at 1 19 30 PM

Comment (kaylahrose) [inline templates/job-form.html:97]: Can you please add spacing between the button and the text on smaller screen sizes?
Screenshot 2025-02-28 at 1 12 37 PM

Comment (kaylahrose) [inline templates/job-form.html:97]: Checkbox is not keyboard accessible, not sure if that is blocking though.

Comment (kaylahrose) [inline templates/job-form.html:97]: Can we make the checkbox required?

File: templates/job-form.html [hunk 4]
```diff
@@ -93,6 +93,20 @@ <h3>Job Listing Form</h3>
...
+              title="Code of Conduct"
+              target="_blank"
+              rel="noopener noreferrer"
+              >
+              Techtonica's Code of Conduct
+            </a>
+            by submitting this form.
+          </label>
+        </div>
         <button id="card-button" type="button" class="sponsor">
```
Comment (kaylahrose) [inline templates/job-form.html:110]: Form validation is not working on submit for me

**Review by kaylahrose — DISMISSED** [review]
Some questions and change requests 🥳

**Comment by daaimah123** [comment]
@kaylahrose Addressed (most-all of) your comments in [e570ac6](https://github.com/Techtonica/techtonica.org/pull/621/commits/e570ac6854294108c11e41f8eda5c4aff876e5fc), [9ee884c](https://github.com/Techtonica/techtonica.org/pull/621/commits/9ee884c0a60dbe55cac2393bb138678103002927), and [7107bbf](https://github.com/Techtonica/techtonica.org/pull/621/commits/7107bbf32e9d37826f842adc77a7b9599a7ae1cb)
- Can you share what you would like to see for accessibility of the checkbox? 
- Alternatively my styling, removing the underline and changing the text to blue doesn't seem to be working beyond adding an additional class; do you all see what I'm doing wrong?

File: templates/job-form.html (hunk 4 above)
Comment (daaimah123) [reply templates/job-form.html:108]: [See comment.](https://github.com/Techtonica/techtonica.org/pull/621#issuecomment-2695782638)

File: templates/job-form.html (hunk 3 above)
Comment (daaimah123) [reply templates/job-form.html]: [See comment.](https://github.com/Techtonica/techtonica.org/pull/621#issuecomment-2695782638)

Comment (daaimah123) [reply templates/job-form.html]: [See comment.](https://github.com/Techtonica/techtonica.org/pull/621#issuecomment-2695782638)

Comment (daaimah123) [reply templates/job-form.html]: Addressed [00bfa38](https://github.com/Techtonica/techtonica.org/pull/621/commits/00bfa38f72a2515b1d26ddd61fae60f67a722e94)

Comment (daaimah123) [reply templates/job-form.html]: Addressed in [00bfa38](https://github.com/Techtonica/techtonica.org/pull/621/commits/00bfa38f72a2515b1d26ddd61fae60f67a722e94)

File: templates/job-form.html (hunk 1 above)
Comment (daaimah123) [reply templates/job-form.html]: https://github.com/Techtonica/techtonica.org/pull/621#issuecomment-2695782638

File: templates/job-form.html (hunk 2 above)
Comment (daaimah123) [reply templates/job-form.html]: Done in [7d1a787](https://github.com/Techtonica/techtonica.org/pull/621/commits/7d1a7877bf459fc4031ca389252505b47074bf9c) and [eb693d6](https://github.com/Techtonica/techtonica.org/pull/621/commits/eb693d6f61216b1e4feef2f0e176a1ce79e31b00)

File: templates/job-form.html (hunk 4 above)
Comment (daaimah123) [reply templates/job-form.html:108]: Validation will be handled with [Payment Form Doesn't Clear Payment Info After Successful Submission](https://github.com/Techtonica/techtonica.org/issues/633#top)

**Comment by daaimah123** [comment]
This issue creates and styles the code of conduct checkbox. I've identified the validation should be handled with [Payment Form Doesn't Clear Payment Info After Successful Submission](https://github.com/Techtonica/techtonica.org/issues/633#top)

**Comment by daaimah123** [comment]
Addressed validation and related styling concerns
post a job

File: static/js/payment-flow.js [hunk 5]
```diff
@@ -94,4 +97,14 @@ window.sendSlackNotification = async function() {
...
 
+// code of conduct checkbox functionality
+document.addEventListener("DOMContentLoaded", () => {
+  const termsCheckbox = document.getElementById("code-of-conduct-checkbox")
+  const cardButton = document.getElementById("card-button")
+
+  termsCheckbox.addEventListener("change", function () {
+    cardButton.disabled = !this.checked
+  })
+})
```
[the remark this answers is not in the document]
Comment (daaimah123) [reply static/js/payment-flow.js]: Addressed in [a43efef](https://github.com/Techtonica/techtonica.org/pull/621/commits/a43efef1552bcb6aff8a5d62b00687acafb8fc42)

File: templates/job-form.html (hunk 3 above)
Comment (daaimah123) [reply templates/job-form.html]: Addressed in [a43efef](https://github.com/Techtonica/techtonica.org/pull/621/commits/a43efef1552bcb6aff8a5d62b00687acafb8fc42)

**Comment by daaimah123** [comment]
@monikkaelyse and @MichelleGlauser, addressed in [11ccbb2](https://github.com/Techtonica/techtonica.org/pull/621/commits/11ccbb2e2a782aeaa3bdf1b00d47dcf700e2a206)

post a job successful submission

**Review by MichelleGlauser — APPROVED** [review]

**Comment by daaimah123** [comment]
Handled blocking "Processing payment information" until the CC is entered. I have not addressed false positives with the profanity checking, @monikkaelyse.

post a job

**Review by ChasVanDav — COMMENTED** [review]
@daaimah123 It's working well for me! 
Screenshot 2025-03-25 at 2 03 39 PM
Screenshot 2025-03-25 at 2 03 49 PM
However, I do wonder if it is too strict?
How about refactoring so that it only flags it if it separated by a space or special character? One edge case is that my best friend's name is Harshitha and she is of Indian decent. So, there may be some cases like this
Screenshot 2025-03-25 at 2 04 33 PM
 where someone's real name wouldn't pass this test.

**Review by ChasVanDav — APPROVED** [review]
Great job on handling edge cases like diverse names! This worked for the case of Harshitha! @daaimah123 🚀 
Screenshot 2025-03-27 at 6 12 02 PM
Screenshot 2025-03-27 at 6 12 11 PM
